x-pack/plugins/security/server/lib/authorization/has_privileges.test.js

```diff
@@ -0,0 +1,311 @@
+      get: jest.fn()
+    }),
+    log: jest.fn()
+  };
+
+  const defaultSettings = {
+    'pkg.version': defaultVersion,
+    'xpack.security.rbac.application': defaultApplication,
+    'kibana.index': defaultKibanaIndex,
+  };
+
+  mockServer.config().get.mockImplementation(key => {
+    return key in settings ? settings[key] : defaultSettings[key];
+  });
+
+  mockServer.savedObjects = {
+    types: savedObjectTypes
+  };
+
+  return mockServer;
+};
+
+const mockApplicationPrivilegeResponse = ({ hasAllRequested, privileges, application = defaultApplication, username = '' }) =>{
+  return {
+    username: username,
+    has_all_requested: hasAllRequested,
+    application: {
+      [application]: {
+        [DEFAULT_RESOURCE]: privileges
+      }
+    }
+  };
+};
+
+const mockKibanaIndexPrivilegesResponse = ({ privileges, index = defaultKibanaIndex }) => {
+  return {
+    index: {
+      [index]: privileges
+    }
+  };
+};
+
+const createMockCallWithRequest = (responses) => {
+  const mockCallWithRequest = jest.fn();
+  getClient.mockReturnValue({
+    callWithRequest: mockCallWithRequest
+  });
+
+  for (const response of responses) {
+    mockCallWithRequest.mockImplementationOnce(async () => response);
+  }
+
+  return mockCallWithRequest;
+};
+
+const expectNoDeprecationLogged = (mockServer) => {
+  expect(mockServer.log).not.toHaveBeenCalled();
+};
+
+const expectDeprecationLogged = (mockServer) => {
+  expect(mockServer.log).toHaveBeenCalledWith(['warning', 'deprecated', 'security'], expect.stringContaining('deprecated'));
+};
+
+test(`returns authorized if they have all application privileges`, async () => {
+  const privilege = `action:saved_objects/${savedObjectTypes[0]}/get`;
+  const username = 'foo-username';
+  const mockServer = createMockServer();
+  const mockCallWithRequest = createMockCallWithRequest([
+    mockApplicationPrivilegeResponse({
+      hasAllRequested: true,
+      privileges: {
+        [getVersionPrivilege(defaultVersion)]: true,
+        [getLoginPrivilege()]: true,
+        [privilege]: true,
+      },
+      application: defaultApplication,
+      username,
+    })
+  ]);
+
+  const hasPrivilegesWithRequest = hasPrivilegesWithServer(mockServer);
+  const request = Symbol();
+  const hasPrivileges = hasPrivilegesWithRequest(request);
+  const privileges = [privilege];
+  const result = await hasPrivileges(privileges);
+
+  expectNoDeprecationLogged(mockServer);
+  expect(mockCallWithRequest).toHaveBeenCalledWith(request, 'shield.hasPrivileges', {
+    body: {
+      applications: [{
+        application: defaultApplication,
+        resources: [DEFAULT_RESOURCE],
+        privileges: [
+          getVersionPrivilege(defaultVersion), getLoginPrivilege(), ...privileges
+        ]
+      }]
+    }
+  });
+  expect(result).toEqual({
+    result: HAS_PRIVILEGES_RESULT.AUTHORIZED,
+    username,
+    missing: [],
+  });
+});
+
+test(`returns unauthorized they have only one application privilege`, async () => {
```

Comment:
I think this message could be more clear about what you're trying to test. As its written, it sounds like we will reject any call that only involves a single application privilege, but I think what you're actually trying to test is that it only takes a single missing privilege to reject the call. If that's true, perhaps `returns unauthorized if they're missing even a single application privilege`?